zebra-network/src/peer_set/set.rs

```diff
@@ -995,78 +1004,74 @@ where
         Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.poll_background_errors(cx)?;
-
-        // Update peer statuses
+        // Update service and peer statuses.
+        //
+        // # Correctness
+        //
+        // If we poll these futures in series, then only the last polled future will wake
+        // the task. This can cause hangs when there are no new peers, no newly ready peers,
+        // or no peers in the peer set at all.
+        //
+        // As a temporary fix, we try to pick the most likely waker, and skip wakers that will
+        // never wake.
+
+        // We can tolerate not being woken for these tasks, because we only return `Pending`
+        // when we have no ready peers. Delays in background error checking are acceptable,
+        // and we can't do anything useful with inventory until we have peers. (The oldest
+        // inventory will get automatically dropped.)
+        let _ = self.poll_background_errors(cx)?;
+        let _ = self.inventory_registry.poll_inventory(cx)?;
+
+        // We must be woken if there are new peers, but new peers can be infrequent if our
+        // connection slots are full, or we're connected to all available/useful peers.
+        //
+        // TODO: wake waiting tasks if there are new peers or unready peers become ready (#7858)
         let _ = self.poll_discover(cx)?;
-        self.disconnect_from_outdated_peers();
-        self.inventory_registry.poll_inventory(cx)?;
-        self.poll_unready(cx);
+        // This method preserves the original waker if it has no unready peers and will never wake.
+        // Otherwise, each connected peer should become ready or timeout within a few minutes.
+        //
+        // TODO: drop peers that overload us with inbound messages and never become ready (#7822)
+        let _ = self.poll_unready(cx)?;
+
+        // Correctness: the waker in the context is waiting on peers. If it is replaced later in
+        // the function, the network could hang. We drop the reference here so it can't be used.
+        #[allow(dropping_references)]
+        std::mem::drop(cx);
```

Comment:
Not needed:
https://github.com/ZcashFoundation/zebra/pull/7859/files#r1380722199